### DISCUSSION TITLE
Enable Link-Time Optimization (LTO)

### DISCUSSION THREAD

**Discussion by zamazan4ik** [gleam-lang/gleam#3677] [Ideas & suggestions]
Hi!

I noticed that in the `Cargo.toml` file Link-Time Optimization (LTO) for the project is not enabled. I suggest switching it on since it will reduce the binary size (always a good thing to have) and will likely improve the application's performance.

I suggest enabling LTO only for the Release builds so as not to sacrifice the developers' experience while working on the project since LTO consumes an additional amount of time to finish the compilation routine. If you think that a regular Release build should not be affected by such a change as well, then I suggest adding an additional `dist` or `release-lto` profile where additionally to regular `release` optimizations LTO will also be added. Such a change simplifies life for maintainers and others interested in the project persons who want to build the most performant version of the application. Using ThinLTO should also help to reduce the build-time overhead with LTO. If we enable it on the Cargo profile level, users, who install the application with `cargo install`, will get the LTO-optimized version "automatically". E.g., check `cargo-outdated` Release [profile](https://github.com/kbknapp/cargo-outdated/blob/master/Cargo.toml#L48).

Basically, it can be enabled with the following lines:

```
[profile.release]
lto = true
```

Thank you.

**Comment by lpil** [comment]
Nice idea! I presume the disadvantage is compile time. Are there others?

Is there a way to only enable this on CI?

**Reply by zamazan4ik** [reply]
> I presume the disadvantage is compile time. Are there others?

Sometimes you can meet specific build issues (nothing in common in my experience - they are unique in each situation). Sometimes there are "interesting" compiler bugs but they are quite rare nowadays in the LTO field.

So I would say no - you shouldn't expect issues with LTO except increased build time and the increased memory usage during the compilation process (it could be important for memory-tight CI envs). And even this issue can be mitigated by using Thin LTO instead of Fat LTO - Thin LTO is a bit less efficient than Fat from the performed compiler optimization perspective but has far less build time overhead (btw Google by default uses ThinLTO in their prod env).

> Is there a way to only enable this on CI?

Yes, you can enable it on CI without changing `Cargo.toml` content via `CARGO_PROFILE` [env vars](https://doc.rust-lang.org/cargo/reference/environment-variables.html) - check [this](https://github.com/alpinelinux/aports/blob/b97c84649f3fb4c7031dba9535328a6d313292ac/testing/c2rust/APKBUILD#L50) example.

However, I suggest enabling it in the Cargo profile directly (maybe not a Release one - it's up to you) since in this case, it's much more clear and visible to the end-users who build Gleam binaries locally (or maintain corresponding packages in their envs). An additional good point - binaries installed directly via `cargo install` will be LTO-optimized.

**Comment by richard-viney** [comment]
Tried this locally on a release build on an M2 MacBook Pro and found:

- Build time went from 52s to 122s, a 2.35x increase
- `gleam` binary size went from 17.61 MiB to 13.56 MiB, a 23% decrease
- A compilation performance benchmark of ~100k lines of Gleam code went from 3.69s to 3.58s, a 3% decrease

LTO is _probably_ worth having due to the modest speed improvement and smaller binary size
